### ISSUE TITLE
RET504 breaks code when value is used in finally

### ISSUE THREAD

**Issue by wikti** [astral-sh/ruff#17292]
### Summary

Had problem with rule RET504. Autofix broke code. 

```
def log_out():
    out = ""
    try:
        out = foo()
        return out
    except Exception as e:
        out = str(e)
    finally:
        log(out)
```

Fix made `return foo()`, which broke `log(out)`.

Tested on version `0.11.4`.

Thanks for great project.

**Comment by MichaReiser** [comment]
I think this is simply a false positive because the assignment isn't unnecessary. One possibility is to make the rule a deferred rule and use the semantic model to verify that the return is the only read. Another alternative is to extend the `ReturnsVisitor` to possibly track the symbols used inside of `except` or `finally` blocks.

**Comment by joaoe** [comment]
+1 I was just about to report this.

My case
```
def execute_something()
    result, err, start = (), None, time.monotonic()
    try:
        logging.info("Starting...")
        result = do_some_long_operation()
        return result ## <- RET504 HERE
    except Exception as e:
        err = repr(e)
        raise
    finally:
        end = time.monotonic()
        logging.log(logging.ERROR if err else logging.INFO, "Result %s, elapsed %.3f secs", err if err else f"{len(result)} rows", end - start)
```

**Comment by mxmlnkn** [comment]
I have a encountered this problem in [fusepy](https://github.com/fusepy/fusepy/blob/5d997d6706cc0204e1b3ca679651485a7e7dda49/fuse.py#L1251).
